Add instructions for Android

k0ekk0ek opened the pull request:
This PR adds instructions on how to cross-compile for Android and run RoundTrip in the Android emulator.

eboasson commented:
Nice! I am not going to try following the instructions right now, but they look like it is very straightforward. Making it work without `pthread_attr_setinheritsched` would be trivial, so this basically means it'll probably work just fine on Android 7 without significant effort.

I'll give it a few days to see if anyone reports running into any difficulties.

joespeed commented:
@k0ekk0ek very promising! Would be **thrilled** by cyclonedds on polestar 2 android emulator https://developer.polestar.com

Help please with these hiccups: [android.cyclonedds_build_errors.txt](https://github.com/eclipse-cyclonedds/cyclonedds/files/4578480/android.cyclonedds_build_errors.txt)

macOS 10.15.4, cmake version 3.17.0, android studio 3.6.3
clang --version
Apple clang version 11.0.3 (clang-1103.0.32.59)
Target: x86_64-apple-darwin19.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

k0ekk0ek commented:
Hi @joespeed! I got the to build errors related to `ddsrt_netstat_*` too. You should be good with the quick-fix to `src/ddsrt/CMakeLists.txt` in this PR, if not, another target system is advertised in CMake. If that's the case, please let me know what the values of the `ANDROID` and `CMAKE_SYSTEM_NAME` variables are. You can do that by putting the following lines in the `src/ddsrt/CMakeLists.txt`:
```
message(STATUS "CMAKE_SYSTEM_NAME: ${CMAKE_SYSTEM_NAME}")
message(STATUS "ANDROID: ${ANDROID}")
```

4.1 Canary seems new enough to have API level 28, but if it's not we can simply come up with replacements for the functions mentioned in the README.md and we should be compatible with most versions of Android.

joespeed commented:
much better!  Many warnings but only 1 error. ~/cyclonedds/build.android
CMake Error: The source directory "/Users/joespeed/Documents" does not appear to contain CMakeLists.txt.

[cyclonedds-android_build.txt](https://github.com/eclipse-cyclonedds/cyclonedds/files/4581729/cyclonedds-android_build.txt)

joespeed commented on ports/android/README.md:
````diff
@@ -0,0 +1,149 @@
+
+ * Push the `RoundTripPong` binary and open a shell.
+```
+$ <path/to/Android/Sdk>/platform-tools push <path/to/cyclonedds>/build.android/bin/RoundTripPong /data/local/tmp
````
```suggestion
$ <path/to/Android/Sdk>/platform-tools/adb push <path/to/cyclonedds>/build.android/bin/* /data/local/tmp
```

joespeed commented on ports/android/README.md:
```diff
@@ -0,0 +1,149 @@
+
+## Deploying and running the RoundTrip example
+
+ * Push the `RoundTripPong` binary and open a shell.
```
```suggestion
 * Push the cyclonedds binaries to Android and open a shell.
```

joespeed commented on ports/android/README.md:
````diff
@@ -0,0 +1,149 @@
+ * Make the binary executable (from the shell on the device).
+```
+$ cd /data/local/tmp
+$ chmod 755 RoundTripPong
````
```suggestion
$ chmod 755 *
```

joespeed commented on ports/android/README.md:
````diff
@@ -0,0 +1,149 @@
+$ export CYCLONEDDS_URI='<CycloneDDS><Domain><General><NetworkInterfaceAddress>eth0</NetworkInterfaceAddress></General></Domain></CycloneDDS>'
+```
+
+ * Execute `$ ./RoundTripPong`
````
```suggestion
 * Execute `$ ./ddsperf -D10 pub & ./ddsperf -D10 sub `
```

joespeed commented on ports/android/README.md:
````diff
@@ -0,0 +1,149 @@
+```
+
+ * Execute `$ ./RoundTripPong`
+ * Build Eclipse Cyclone DDS for your host machine
````
remove the sections on building for host & setting up networking OR move them to the end as fyi

joespeed commented on ports/android/README.md:
```diff
@@ -0,0 +1,149 @@
+
+ * Execute `$ ./RoundTripPong`
+ * Build Eclipse Cyclone DDS for your host machine
+ * Force the host to use the `br0` interface
```
remove or move to fyi section

joespeed commented on ports/android/README.md:
````diff
@@ -0,0 +1,149 @@
+$ chmod 755 RoundTripPong
+```
+
+ * Force the guest to use the `eth0` interface.
````
remove or move to fyi section at end

joespeed commented on ports/android/README.md:
```diff
@@ -0,0 +1,149 @@
+> is mounted with the `noexec` flag and will not allow you to execute any
+> executables.
+
+ * Make the binary executable (from the shell on the device).
```
```suggestion
 * Make cyclonedds binaries executable (from the shell on the device).
```

joespeed commented on ports/android/README.md:
```diff
@@ -0,0 +1,149 @@
+> for demonstration purposes.
+
+
+## Setting up a networking bridge
```
remove network setup or move to fyi section at end

k0ekk0ek replied:
The network setup is actually the most interesting part because it contains the bare minimum of steps to get things working. Makes debugging etc a lot easier, so a people likely want to have this information once they get going. Apart from that it shows that it'll communicate over the network, not just `lo`. I can move it to the end and add a local test step in between.

k0ekk0ek commented:
Thanks for reviewing @joespeed! I changed the example to use `ddsperf` instead and moved the instructions for setting up a bridge to the end. I purposely did not change the commands to copy all binaries and make them executable using `chmod 755 *` because people might use these same instructions on an actual device. Shouldn't be to big of a problem as most people will only use these instructions to verify it works. Afterwards they're likely to copy their own application. At some point we should add decent instructions on how to link against Eclipse Cyclone DDS from an Android Studio project, or even better, add a Java binding too. For now, I'm positive this'll suffice(?)
